AK: Do not VERIFY while iterating invalid code point bytes in UTF8View, return a U+FFFD 'REPLACEMENT CHARACTER' instead

Dexesttp opened the pull request:
# Overview

This PR makes it so we output the 0xFFFD 'REPLACEMENT CHARACTER' code point when encountering invalid bytes in an `UTF8CodepointIterator`, and keep iterating the view after skipping one byte.

The previous behavior was to always `VERIFY` that the UTF-8 bytes were valid, causing crashes every time a string was read prior to checking its validity with `UTF8View::validate()`. This caused crashes in applications (#5896) and was encountered by fuzzers (for example, [on `LibJS`](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=29252&sort=-id&q=proj%3Dserenity)).

## Alternate Implementations

### Expect the callers to validate the view before iterating on it

This was the current implementation so far, but it is not only cumbersome but inefficient. Moreover, there was no way for callers to easily iterate on correct part of the strings : the only option was to reject the input altogether.

As this still caused issues in many parts of the code, this retrospectively doesn't seem like the best way to do this. The `AK::UTF8CodepointIterator` being more lenient will give more flexibility to the implementations, with a fallback in case they don't want to handle this complexity.

### Return an `Optional<u32>` and let the callers manually handle invalid code points.

One alternative would be to stop returning an `u32` and to instead return an `Optional<u32>`, and then leave the decision to the consumer whether to replace the code point or to stop iterating the string. This requires more extensive changes all over the codebase to handle this new behavior, but isn't hard by itself.

An attempt was made here: [DexesTTP/optional-utf8](https://github.com/Dexesttp/serenity/tree/tasks/optional-utf8)

However, this alternative breaks symmetry with the `UTF32View` and `StringView` API, which would require either heavy refactoring and/or code duplication in the currently templated code found in [`Gfx::Painter::draw_text_line`](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGfx/Painter.cpp#L1200) and generic lambda found in [`Shell/find_offset_into_node()`](https://github.com/SerenityOS/serenity/blob/master/Userland/Shell/Shell.h#L367).

### Skip the estimated code point length instead of one single byte

The current implementation skips one single byte, therefore returning possibly several followup replacement characters in case one single multi-byte invalid code point is encountered.

The UTF-8 specification would only encounter this kind of behavior at the very end of the string (e.g. where a 3-byte code point is expected, but only 2 bytes are available). Our implementation would return two `U+FFFD` glyphs in this case : one for the first byte, and one for the second byte which couldn't be decoded directly.

Our current implementation allows to distinguish between a truncated glyph (outputed as two replacement characters) and a corrupted byte in the non-final position (outputed as one replacement character and the normal character for the following byte). As such, this seemed like the better implementation.

BenWiederhake commented:
This will likely fix https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=29252&sort=-id&q=proj%3Dserenity

MaxWipfli commented on AK/Utf8View.cpp:
```diff
@@ -212,6 +226,11 @@ size_t Utf8CodepointIterator::code_point_length_in_bytes() const
+
+    if (!first_byte_makes_sense) {
+        // We will output a replacement character for this byte and treat it as a codepoint of size 1.
+        return 1;
```
We have a problem here:
Does this function return the length of the data as stored inside the view or as it will be returned by `operator*`?
If the former, returning 1 is correct. If the latter, we need to return 3 (as U+FFFD is encoded as `0xef 0xbf 0xbd`). See the review comment for more thoughts about this.

Dexesttp replied:
I don't think it's as much of an issue as you make it seem (reasons explained below), but I agree that it could be clearer. I updated the name of the method from `code_point_length_in_bytes()` to `underlying_code_point_length_in_bytes()` and added a comment in the header explaining this behavior.

MaxWipfli commented on AK/Utf8View.cpp:
```diff
@@ -224,14 +243,27 @@ u32 Utf8CodepointIterator::operator*() const
         dbgln("First byte doesn't make sense, bytes: {}", StringView { (const char*)m_ptr, m_length });
+        return 0xFFFD;
+    }
     VERIFY(first_byte_makes_sense);
```
You do not need to `VERIFY` this because if it is false, the `if` above would have already returned.

Dexesttp replied:
The VERIFY still makes sense : at this point of the code, we want to be sure that the first byte makes sense otherwise the code is invalid.

Rather than reaching an unrelated crash if somebody (somehow) changes the condition and we break an implicit VERIFY, I would rather make it explicit even if redundant and let the compiler optimize it away if needed !

MaxWipfli replied:
I see what you mean, but this argument could be made for every `if` statement that returns early then. But I guess that's a question of preference, so let's keep it that way.

MaxWipfli commented on AK/Utf8View.cpp:
```diff
@@ -224,14 +243,27 @@ u32 Utf8CodepointIterator::operator*() const
         dbgln("Not enough bytes (need {}, have {}), first byte is: {:#02x}, '{}'", code_point_length_in_bytes, m_length, m_ptr[0], (const char*)m_ptr);
+        return 0xFFFD;
+    }
     VERIFY(code_point_length_in_bytes <= m_length);
```
Same here.

Dexesttp replied:
Same as above : this VERIFY is valid, even if trivial. The compiler will optimize it away, but it will defend against future code.

MaxWipfli commented on AK/Utf8View.cpp:
```diff
@@ -224,14 +243,27 @@ u32 Utf8CodepointIterator::operator*() const
+            dbgln("Extension byte {:#02x} in {} position after first byte {:#02x} doesn't make sense.", m_ptr[offset], offset, m_ptr[0]);
+            return 0xFFFD;
+        }
         VERIFY(m_ptr[offset] >> 6 == 2);
```
Same here.

Dexesttp replied:
Same as above : this VERIFY is valid, even if trivial. The compiler will optimize it away, but it will defend against future canges to the code.

MaxWipfli reviewed (changes requested):
I have been working on and with `Utf8View` quite a bit and I think your changes probably make sense. Please change what I commented in the code.

There is one problem though:  
We have to keep in mind that we now have a discrepancy between what the bytes inside the `Utf8View` represent and what you get when using a `StringBuilder` and appending all codepoints by iterating over the view. We have to keep in mind that this could lead to some weird bugs, although I don't see what we could do about this (except "parsing" the data on construction and actually replacing invalid bytes with U+FFFD, but this would require a copy of the data (which is not what a "view" tries to achieve).

When operating with strings on the byte level, small mistakes can become basically impossible to track down. Using `VERIFY` to fail as soon as possible and as loud as possible thus makes sense (in a non-production environment as we have it now). With this patch, we introduce the following problem:

**Situation 1**  
You are iterating over a Unicode string and want to allocate some bytes to store your data. You expect `Utf8CodepointIterator::code_point_length_in_bytes()` to return the byte length of the codepoint that you get with `operator*`.

**Situation 2**  
You are doing some offset math to determine what byte offset a certain codepoint has in the original string (see `Utf8View::iterator_at_byte_offset(size_t)`). You expect `code_point_length_in_bytes` to return how wide the iterator's codepoint is in the original string.

These to situations require different return values. Yes. we could add a second function to cover the other case, but then it starts to get complicated and we should start thinking about naming.

I understand why you want to change this, and I do think we eventually have to find a solution so programs stop crashing on random garbage user input, but I think this requires some more thought/discussion (and some documentation in the form of comments, which will explain to the user what he should expect from `Utf8CodepointIterator`).

I am quite sure the code as it stands will introduce a bug somewhere due to this discrepancy between the underlying memory and what happens when iterating over it.

Dexesttp commented:
I think your concerns are valid, but are kind of different to the current behavior of the `UTF8CodepointIterator`.

The iterator can currently be seen as "iterate over the given UTF-8 string as if it was a UTF-32 string" - it returns a series of u32 codepoints.

Anybody wanting to actually re-encode the UTF-8 string based on the _original_ UTF-8 bytes currently uses the underlying bytes directly (as is done in `TextNode::compute_text_for_rendering`). The other parts of the code store the `UTF-32`-decoded glyphs in an `UTF-32`-style storage such as a `Vector<u32>` (as is done in `Painter/split_text_into_directional_runs`).

The need to iterate over an `UTF-8` string and then reencode the contents still in `UTF-8` based on the re-encoded `UTF-32`-decoded glyph, while you have the raw data right there, seems very... Incorrect. The fact that this would cause undefined behavior wouldn't be surprising.

MaxWipfli commented:
> The iterator can currently be seen as "iterate over the given UTF-8 string as if it was a UTF-32 string" - it returns a series of u32 codepoints.

Didn't think about it this way, makes sense.

> Anybody wanting to actually re-encode the UTF-8 string based on the _original_ UTF-8 bytes currently uses the underlying bytes directly (as is done in `TextNode::compute_text_for_rendering`). The other parts of the code store the `UTF-32`-decoded glyphs in an `UTF-32`-style storage such as a `Vector<u32>` (as is done in `Painter/split_text_into_directional_runs`).

I actually think this code just wants to use `StringBuilder::apend_code_point`, but it didn't for some reason. Are do you know why this is needed this way? This should probably be changed.

> The need to iterate over an `UTF-8` string and then reencode the contents still in `UTF-8` based on the re-encoded `UTF-32`-decoded glyph, while you have the raw data right there, seems very... Incorrect. The fact that this would cause undefined behavior wouldn't be surprising.

Yes, makes sense.

But according to your thinking, `code_point_length_in_bytes` needs to return 3, not 1, as it should be the length of the code point you get when using `operator*`.

Furthermore, we should probably have a method `bool Utf8CodepointIterator::is_valid_codepoint()` that tells the caller if the code point has been replaced by U+FFFD.

I will test the current usage of `Utf8View::iterator_at_byte_offset()` with your patch to see if that requires modification (I think so).

MaxWipfli commented:
Everything seems fine and ready to be merged. My concerns have been nullified by @Dexesttp. :^)

Dexesttp commented:
Some small updates :
- Replaced comments talking about "codepoints" with "code points" instead (since that's what we're going for for now on)
- Added a few tests for the new behavior

Dexesttp commented:
Some more updates:
- Rebased because it didn't compile anymore
- Added some more tests to make sure it really works (it does)
- Added the `underlying_code_point_bytes()` method, and immediately use it in the one place where it's needed, as discussed on Discord

alimpfard commented:
Let's merge this, the code looks good and the CI is happy.
